File: templates/code/python/github_mcp_client.py
"""GitHub MCP Client Template

Reusable client for interacting with GitHub via MCP tools.

Usage:
    client = GitHubMCPClient()
    issues = await client.list_issues('owner', 'repo', state='open')
"""

import asyncio
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta
import json


class RateLimiter:
    """Rate limiter with sliding window"""
    
    def __init__(self, per_minute: int = 20, per_hour: int = 120):
        self.per_minute = per_minute
        self.per_hour = per_hour
        self.minute_count = 0
        self.hour_count = 0
        self.minute_window = int(datetime.now().timestamp() // 60)
        self.hour_window = int(datetime.now().timestamp() // 3600)
    
    async def check_and_wait(self) -> None:
        """Check limits and wait if necessary"""
        now = datetime.now()
        current_minute = int(now.timestamp() // 60)
        current_hour = int(now.timestamp() // 3600)
        
        # Reset counters on window change
        if current_minute != self.minute_window:
            self.minute_count = 0
            self.minute_window = current_minute
        
        if current_hour != self.hour_window:
            self.hour_count = 0
            self.hour_window = current_hour
        
        # Wait if limits exceeded
        if self.minute_count >= self.per_minute:
            wait_seconds = 60 - (now.timestamp() % 60)
            print(f"Minute rate limit reached. Waiting {wait_seconds:.1f}s...")
            await asyncio.sleep(wait_seconds)
            self.minute_count = 0
        
        if self.hour_count >= self.per_hour:
            wait_seconds = 3600 - (now.timestamp() % 3600)
            print(f"Hour rate limit reached. Waiting {wait_seconds:.1f}s...")
            await asyncio.sleep(wait_seconds)
            self.hour_count = 0
        
        # Increment counters
        self.minute_count += 1
        self.hour_count += 1


class GitHubMCPClient:
    """High-level GitHub MCP client with rate limiting"""
    
    def __init__(self, rate_limit: bool = True):
        self.rate_limiter = RateLimiter() if rate_limit else None
    
    async def _call_tool(self, tool_name: str, **params) -> Dict[str, Any]:
        """Call MCP tool with rate limiting"""
        if self.rate_limiter:
            await self.rate_limiter.check_and_wait()
        
        # In real implementation, this would call the actual MCP tool
        # For template purposes, this is a placeholder
        raise NotImplementedError("Implement tool calling mechanism")
    
    async def list_issues(
        self,
        owner: str,
        repo: str,
        state: str = "open",
        per_page: int = 30,
        max_results: Optional[int] = None
    ) -> List[Dict[str, Any]]:
        """List issues with automatic pagination"""
        issues = []
        after = None
        
        while True:
            result = await self._call_tool(
                "mcp_tool_github-mcp-direct_list_issues",
                owner=owner,
                repo=repo,
                state=state,
                perPage=per_page,
                after=after
            )
            
            issues.extend(result.get("issues", []))
            
            # Check pagination
            page_info = result.get("pageInfo", {})
            if not page_info.get("hasNextPage"):
                break
            
            after = page_info.get("endCursor")
            
            # Check max results
            if max_results and len(issues) >= max_results:
                issues = issues[:max_results]
                break
        
        return issues
    
    async def create_issue(
        self,
        owner: str,
        repo: str,
        title: str,
        body: str,
        labels: Optional[List[str]] = None,
        assignees: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """Create a new issue"""
        return await self._call_tool(
            "mcp_tool_github-mcp-direct_issue_write",
            method="create",
            owner=owner,
            repo=repo,
            title=title,
            body=body,
            labels=labels or [],
            assignees=assignees or []
        )
    
    async def get_file_contents(
        self,
        owner: str,
        repo: str,
        path: str,
        ref: Optional[str] = None
    ) -> str:
        """Get file contents from repository"""
        result = await self._call_tool(
            "mcp_tool_github-mcp-direct_get_file_contents",
            owner=owner,
            repo=repo,
            path=path,
            ref=ref or ""
        )
        return result.get("content", "")


async def main():
    """Example usage"""
    client = GitHubMCPClient()
    
    # List open issues
    issues = await client.list_issues(
        owner="pv-udpv",
        repo="ai-agent-research",
        state="open",
        max_results=10
    )
    print(f"Found {len(issues)} open issues")
    
    # Create a new issue
    new_issue = await client.create_issue(
        owner="pv-udpv",
        repo="ai-agent-research",
        title="Research Task: Investigate Tool X",
        body="Detailed description...",
        labels=["RESEARCH"]
    )
    print(f"Created issue #{new_issue['number']}")


if __name__ == "__main__":
    asyncio.run(main())

File: templates/code/typescript/github-mcp-client.ts
/**
 * GitHub MCP Client Template (TypeScript)
 * 
 * Reusable client for GitHub operations with rate limiting.
 * 
 * Usage:
 *   const client = new GitHubMCPClient();
 *   const issues = await client.listIssues('owner', 'repo');
 */

interface RateLimitConfig {
  perMinute: number;
  perHour: number;
}

interface Issue {
  id: string;
  number: number;
  title: string;
  body: string;
  state: string;
  labels: string[];
  created_at: string;
  updated_at: string;
}

interface PageInfo {
  hasNextPage: boolean;
  endCursor?: string;
}

class RateLimiter {
  private perMinute: number;
  private perHour: number;
  private minuteCount: number = 0;
  private hourCount: number = 0;
  private minuteWindow: number;
  private hourWindow: number;

  constructor(config: RateLimitConfig = { perMinute: 20, perHour: 120 }) {
    this.perMinute = config.perMinute;
    this.perHour = config.perHour;
    this.minuteWindow = Math.floor(Date.now() / 60000);
    this.hourWindow = Math.floor(Date.now() / 3600000);
  }

  async checkAndWait(): Promise<void> {
    const now = Date.now();
    const currentMinute = Math.floor(now / 60000);
    const currentHour = Math.floor(now / 3600000);

    // Reset counters
    if (currentMinute !== this.minuteWindow) {
      this.minuteCount = 0;
      this.minuteWindow = currentMinute;
    }

    if (currentHour !== this.hourWindow) {
      this.hourCount = 0;
      this.hourWindow = currentHour;
    }

    // Wait if needed
    if (this.minuteCount >= this.perMinute) {
      const waitMs = 60000 - (now % 60000);
      console.log(`Minute rate limit reached. Waiting ${waitMs}ms...`);
      await new Promise(resolve => setTimeout(resolve, waitMs));
      this.minuteCount = 0;
    }

    if (this.hourCount >= this.perHour) {
      const waitMs = 3600000 - (now % 3600000);
      console.log(`Hour rate limit reached. Waiting ${waitMs}ms...`);
      await new Promise(resolve => setTimeout(resolve, waitMs));
      this.hourCount = 0;
    }

    // Increment
    this.minuteCount++;
    this.hourCount++;
  }
}

export class GitHubMCPClient {
  private rateLimiter?: RateLimiter;

  constructor(rateLimit: boolean = true) {
    this.rateLimiter = rateLimit ? new RateLimiter() : undefined;
  }

  private async callTool<T>(
    toolName: string,
    params: Record<string, any>
  ): Promise<T> {
    if (this.rateLimiter) {
      await this.rateLimiter.checkAndWait();
    }

    // In real implementation, call actual MCP tool
    throw new Error("Implement tool calling mechanism");
  }

  async listIssues(
    owner: string,
    repo: string,
    options: {
      state?: "open" | "closed";
      perPage?: number;
      maxResults?: number;
    } = {}
  ): Promise<Issue[]> {
    const { state = "open", perPage = 30, maxResults } = options;
    const issues: Issue[] = [];
    let after: string | undefined;

    while (true) {
      const result = await this.callTool<{
        issues: Issue[];
        pageInfo: PageInfo;
      }>("mcp_tool_github-mcp-direct_list_issues", {
        owner,
        repo,
        state,
        perPage,
        after,
      });

      issues.push(...result.issues);

      if (!result.pageInfo.hasNextPage) break;
      after = result.pageInfo.endCursor;

      if (maxResults && issues.length >= maxResults) {
        return issues.slice(0, maxResults);
      }
    }

    return issues;
  }

  async createIssue(
    owner: string,
    repo: string,
    data: {
      title: string;
      body: string;
      labels?: string[];
      assignees?: string[];
    }
  ): Promise<Issue> {
    return this.callTool<Issue>("mcp_tool_github-mcp-direct_issue_write", {
      method: "create",
      owner,
      repo,
      ...data,
      labels: data.labels || [],
      assignees: data.assignees || [],
    });
  }

  async getFileContents(
    owner: string,
    repo: string,
    path: string,
    ref?: string
  ): Promise<string> {
    const result = await this.callTool<{ content: string }>(
      "mcp_tool_github-mcp-direct_get_file_contents",
      { owner, repo, path, ref: ref || "" }
    );
    return result.content;
  }
}

// Example usage
async function main() {
  const client = new GitHubMCPClient();

  // List issues
  const issues = await client.listIssues("pv-udpv", "ai-agent-research", {
    state: "open",
    maxResults: 10,
  });
  console.log(`Found ${issues.length} issues`);

  // Create issue
  const newIssue = await client.createIssue("pv-udpv", "ai-agent-research", {
    title: "Research Task: Investigate Tool X",
    body: "Detailed description...",
    labels: ["RESEARCH"],
  });
  console.log(`Created issue #${newIssue.number}`);
}
